Fixed #30899 -- Lazily compiled import time regular expressions.

hramezani opened the pull request:
[Ticket 30899](https://code.djangoproject.com/ticket/30899)

- Just import time `re.compile` are changed.
- Tests are excluded.
- [django.utils.translation.trim_whitespace_re](https://github.com/django/django/blob/6bbf9a20e2c5865e01b537e8cd34dfca06621a4a/django/utils/translation/__init__.py#L331) did not include because `django.utils.translation` already imported in `django.utils.text`.(Circular dependency)
- [django.core.handlers.wsgi._slashes_re](https://github.com/django/django/blob/6bbf9a20e2c5865e01b537e8cd34dfca06621a4a/django/core/handlers/wsgi.py#L12) did not include because it compile bytes `re.compile(br'/+')` and `_lazy_re_compile` does not handle bytes(add byte handling to `_lazy_re_compile`?)

adamchainz commented:
> django.utils.translation.trim_whitespace_re did not include because django.utils.translation already imported in django.utils.text.(Circular dependency)

Perhaps we should put the utility in `django.utils.regex_helper` instead, if that fixes it.

> .django.core.handlers.wsgi._slashes_re did not include because it compile bytes re.compile(br'/+') and _lazy_re_compile does not handle bytes(add byte handling to _lazy_re_compile?)

Please add byte handling - seems like it would be straightforward!

hramezani commented:
@adamchainz Thanks for the comment.
I moved `_lazy_re_compile` to `django.utils.regex_helper`, and changed below line to use `_lazy_re_compile` : https://github.com/django/django/blob/6bbf9a20e2c5865e01b537e8cd34dfca06621a4a/django/utils/translation/__init__.py#L331
Also, Added bytes handling to `_lazy_re_compile`  and changed the below line to use `_lazy_re_compile` :
https://github.com/django/django/blob/6bbf9a20e2c5865e01b537e8cd34dfca06621a4a/django/core/handlers/wsgi.py#L12

pope1ni commented:
I note from the ticket that this change is intended to reduce delays during import.

Could you show an example from before and after using [`-X importtime`](https://docs.python.org/3/using/cmdline.html#id5)? Probably something like `python -X importtime -m django runserver` on a project that imports some of these regex-containing modules.

hramezani commented:
@pope1ni Thanks for the comment. 
I ran the test on master against this patch on my local machine. 
Here is the report for some changed modules:

```
import time: self [us]  | cumulative | imported package
------------------------|------------|-----------------------------
import time:      2861  | 3181       | django.utils.text (before)
import time:       491  | 824        | django.utils.text (after)
import time:      2392  |       2392 | django.utils.dateparse (before)
import time:      123   |       123  | django.utils.dateparse (after)
import time:      2743  |       8098 | django.utils.html (before)
import time:      448   |       3467 | django.utils.html (after)
```

pope1ni commented:
Wow. That looks like a much bigger delta than I would have expected.

Did you run multiple iterations of before and after to rule out any I/O?

Otherwise, this is neat. :sunglasses:

adamchainz commented:
Fantastic! "A few milliseconds ✖️100 times a day ✖️every django developer from now on" should come out pretty big :)

felixxm commented:
@hramezani Thanks, great job :rocket:
